Work on migrating to conan v2

php1ic opened the pull request:
Updates on the road to getting things working with conan v2.

This should **NOT** yet be merged as it doesn't compile. There are include errors relating to 1 of the packages pulled in by conan (nlohmann) but I can't see why that package is any different to the others so need some help.

php1ic committed:
Update conanfile to work with version 2

The way generators are handled has changed so update that section

Changes in conan 2 mean packages *must* be all lowercae and the
Qt-Color-Widget package has been updated so this project needs to track
the change.

Update option string to work with conan v2.

php1ic committed:
Update top level CMakeLists file and README with config instructions

Conan v2 has changed the way it interacts with CMake. At the moment, you
need to clone the cmake-conan repo separately and point to it's .cmake
file when configuring your project.

Perhaps the cmake-conan repo should be added as a submodule of the main
project?

ferdymercury commented on README.md:
```diff
@@ -41,12 +41,15 @@ conan remote add ess-dmsc https://artifactory.esss.lu.se/artifactory/api/conan/e
 The ESS remote is required for QtColorWidgets. In case of changes to their deployment infrastructure, check for updates [here](https://github.com/ess-dmsc/conan-configuration).
 
+Changes in conan v2 require the cloning of the [cmake-conan](https://github.com/conan-io/cmake-conan/tree/develop2) repository, and use of the develop2 branch.
+Once the repository is cloned and the develop2 branch checked out, this repo can be configured as shown.
```
I think it would be great to specify a oneliner command doing the git clone of already the specific branch.

php1ic replied:
How about adding the cmake-conan repo as a submodule? We can then set the cmake option in the CMakeLists file as we know the full path to the necessary file.

ferdymercury replied:
Sounds good! We could also try FetchContent instead of a submodule? https://cmake.org/pipermail/cmake/2019-June/069709.html

martukas replied:
I would to not have either submodule or assumption of anything cloned or installed to the system that does not come through a common package manager such as apt/pip/brew. From my perspective `cmake-conan` loses its utility for conan2, the way they have made it.
How about just having the user run the conan command manually prior to building? That's one extra line, but that's still less than submodule shenanigans.

ferdymercury replied:
Thanks php1ic.  I tried your branch, however I cannot compile on Ubuntu 22, it says:

```
======== Finalizing install (deploy, generators) ========
conanfile.txt: Writing generators to /tmp/nuclei/build/build/conan
conanfile.txt: Generator 'CMakeDeps' calling 'generate()'
conanfile.txt: Generator 'CMakeToolchain' calling 'generate()'
conanfile.txt: CMakeToolchain generated: conan_toolchain.cmake
conanfile.txt: Preset 'conan-release' added to CMakePresets.json. Invoke it manually using 'cmake --preset conan-release' if using CMake>=3.23
conanfile.txt: If your CMake version is not compatible with CMakePresets (<3.23) call cmake like: 'cmake <path> -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE=/tmp/nuclei/build/build/conan/conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release'
conanfile.txt: CMakeToolchain generated: CMakePresets.json
conanfile.txt: CMakeToolchain generated: ../../../CMakeUserPresets.json
conanfile.txt: Generating aggregated env files
conanfile.txt: Generated aggregated env files: ['conanbuild.sh', 'conanrun.sh']
Install finished successfully
-- CMake-Conan: CONAN_GENERATORS_FOLDER=/tmp/nuclei/build/build/conan
-- CMake-Conan: CONANFILE=/tmp/nuclei/conanfile.txt
CMake Error at /tmp/cmake-conan/conan_provider.cmake:278 (find_package):
  Could not find a package configuration file provided by "Threads" with any
  of the following names:

    ThreadsConfig.cmake
    threads-config.cmake

  Add the installation prefix of "Threads" to CMAKE_PREFIX_PATH or set
  "Threads_DIR" to a directory containing one of the above files.  If
  "Threads" provides a separate development package or SDK, be sure it has
  been installed.
Call Stack (most recent call first):
  CMakeLists.txt:7 (find_package)

```

ferdymercury replied:
> I would to not have either submodule or assumption of anything cloned or installed to the system that does not come through a common package manager such as apt/pip/brew. From my perspective `cmake-conan` loses its utility for conan2, the way they have made it. How about just having the user run the conan command manually prior to building? That's one extra line, but that's still less than submodule shenanigans.

Sounds good to me. What would be then the exact install command for conan?

php1ic replied:
This configuration issue with the `Threads` package could be another reason to move away from `cmake-conan`. When I made the PR the tip of the `develop2` branch was at [38a78aaf](https://github.com/conan-io/cmake-conan/commit/38a78aaf3a5567bd29932bf23e9) but the tip now also gives me that error.

Running conan externally, then cmake, does not give this issue. I'll update the README with the conan command to run and modified cmake commands.

CMake introduced [presets](https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html) to simply configuration and conan 2 makes use of that. It arrived in cmake 3.19, the newest release is 3.26 and the latest [ubuntu LTS](https://packages.ubuntu.com/search?keywords=cmake) is on 3.22. Any issues with me using presets in the instructions?

ferdymercury replied:
Good idea the presets, thanks!

php1ic replied:
I should have read the conan output more carefully. Presets arrived in 3.19, but conan uses functionality from 3.23 so they can't be used by everyone at the moment.

The cmake command looks worse than it is, but if we want to make it super easy, we could write a simple build script?

martukas replied:
Where is the dividing line for who has the new CMake? I don't see any need to actively support anything older than what the latest Ubuntu LTS has.

ferdymercury replied:
Ubuntu LTS has 3.22 but conan uses 3.23, i think that was the issue

php1ic replied:
Correct, 3.23 introduced functionality that conan makes use of and the Ubuntu LTS (and thus the CI image) is on 3.22

ferdymercury commented:
Maybe this helps?
https://github.com/nlohmann/json/discussions/3483
https://json.nlohmann.me/integration/package_managers/#conan

php1ic committed:
Specify all the packages that are used

This is needed to get the project compiling. Perhaps the method of
calling conan1 automagically found and included everything. This does
not seem to be the case with conan2 so we have to manaully add.

php1ic committed:
Add the cstdint include for uint16_t

This was required to compile with gcc 13.1.1

php1ic committed:
Ignore more autogenerated files

Conan now creates a preset file that can be used with newer versions of
cmake.

IDE like feature with some editors (e.g. neovim) work best when you get
cmake to create a `compile_commands.json` file that can be used by lsp.
This then creates a `.cache` file where is does it's work. We shouldn't
track either.

php1ic commented:
The branch now compiles and I can run the program :+1: 

In the top level CMakeLists file I had to explicitly call `find_package()` on all the packages installed via conan then make sure they were linked against when building the main executable. Perhaps in conan v1 a lot of that was automagically done by the code used in cmake and the file it downloaded?

martukas commented on .gitignore:
```diff
@@ -14,3 +14,11 @@ cmake-build-*/
+CMakeUserPresets.json
+
+#Compilation database used by lsp
+compile_commands.json
```
what's lsp?

ferdymercury replied:
https://github.com/emacs-lsp/lsp-mode/discussions/3478 maybe?

php1ic replied:
Yep, Language Server Protocol - https://en.wikipedia.org/wiki/Language_Server_Protocol & https://neovim.io/doc/user/lsp.html

In this case the file is used by [clangd](https://clangd.llvm.org/) so that may be a better name to use in the comment.

martukas reviewed (changes requested):
I would update the CI script as well to build it in the new way. It's a good way to make sure it works in general, not just on your machine. It's also blocking the merge.

php1ic committed:
Update build instructions with new method

It's just as simple to run conan and install the packages prior to the
cmake configure stage so lets now do that.

Move to using cmake presets to simply things and increase minimum
required version accordingly.

php1ic committed:
Update circleCI config

Get the latest version of conan from pip, call conan externally prior
to cmake and use the autogenerated preset to configure and build.

php1ic committed:
Update CI after changes to conan in version 2

Creating or detecting a new profile has changed and it is no longer
possible to edit settings from the command line. Luckily for us, it
appears that the setting we wanted to change to is now the default.

https://github.com/conan-io/conan/issues/12833

php1ic committed:
Update CI config and README for older versions of cmake

Although presets arrived with 3.19, conan v2 uses functionality that was
release with 3.23. The current ubuntu LTS, which is used on circleCI,
has 3.22 so we can't use the presets created by conan in that config.

The presets do make config and build simpler, but we must assume that not
all people are on a new enough version of cmake so the README needs to
have instructions for those users.

martukas commented on CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.7.0 FATAL_ERROR)
+cmake_minimum_required(VERSION 3.19.0 FATAL_ERROR)
```
Must it be 19 now? Might be too restrictive? Could it be just blanket 3.anything? Or do we do things that are only available in latest?

martukas commented on README.md:
````diff
@@ -41,12 +41,23 @@ conan remote add ess-dmsc https://artifactory.esss.lu.se/artifactory/api/conan/e
+If your version of cmake is older than 3.23
+```bash
+cd build
+cmake .. -DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release
````
Wow, yeah this does look way more complicated than I'm used to. I understand "Release", but can't the other things just already be defined somewhere in the CMake files?
And this CMP0091 thing also looks like a sign of something funky. Does this have to do anything with accommodating old ABIs? I think the more common approach is to run `conan profile update settings.compiler.libcxx=libstdc++11 default`  after the conan config detect.

ferdymercury replied:
I agree, I would suggest putting inside CMakeLists : if cmake version... Then set policy ..

And maybe also:
```
set(default_build_type "Release")
if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
  message(STATUS "Setting build type to '${default_build_type}' as none was specified.")
  set(CMAKE_BUILD_TYPE "${default_build_type}" CACHE STRING "Choose the type of build." FORCE)
  # Set the possible values of build type for cmake-gui
  set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release")# "MinSizeRel" "RelWithDebInfo"
endif()
```

php1ic replied:
>  I think the more common approach is to run `conan profile update settings.compiler.libcxx=libstdc++11 default` after the conan config detect.

This functionality was [removed](https://github.com/conan-io/conan/issues/12833) in v2

I copied the advice/instruction given by conan once it was run, but I think there like to be explicit, so we can safely drop 2 of the command line options as:
- The build type is checked and set in [EnsureBuildType.cmake](https://github.com/martukas/nuclei/blob/main/cmake/EnsureBuildType.cmake) so a new user doesn't need to explicitly set it.
- The `CMP0091` setting is [MSVC](https://cmake.org/cmake/help/latest/policy/CMP0091.html) specific. Looks like this project only checks for MSVC in the boost config and not in a way related to this setting.

The other option (`CMAKE_TOOLCHAIN_FILE`) tells cmake where the packages are installed so is definitely needs. We could assume the user has copy/pasted the command(s), so we know the paths and can set the option in the top level cmake file.

Another option could be to run conan from inside cmake with [execute_process](https://cmake.org/cmake/help/v3.22/command/execute_process.html), but this feels like we are re-inventing what [cmake-conan](https://github.com/conan-io/cmake-conan) is doing so there must be some intricacies I'm not aware of.

martukas reviewed (changes requested):
Everything looks almost ready. And great to see CI building successfully now.
But the build process still actually looks more complicated than it should be. 
If that truly cannot be avoided then yes, a build scirpt is maybe a good idea. But I do suspect it can be a bit simplified?

php1ic committed:
Move the conan step into cmake to simply configuration

New cmake file wraps up conan functionality and is called from the top
level cmake file. The setting of the toolchain file is done in the top
level file incase we decide to not do the conan step from within cmake.
If so, we can just delete the new file and the call to it.

The README and CI have also been updated to reflect these changes.

php1ic commented:
Conan work has been moved into it's own cmake file and we are back to only needing to run `cmake ..` from the build directory.

Minor annoyance is that conan will always be called with cmake as I couldn't find a good way to check if packages are already installed thus stopping subsequent (unnecessary) calls to conan. If we don't want to do this, we can delete the new cmake file, remove the call from the top level cmake file and have a separate call to conan. The cmake call will still just be `cmake ..` as the variable is set externally to the conan specific file.

ferdymercury commented on .gitignore:
```diff
@@ -14,3 +14,11 @@ cmake-build-*/
+#Conan created files
+CMakeUserPresets.json
+
+#Compilation database used by lsp
```
```suggestion
#Compilation database used by lsp (clangd)
```

ferdymercury commented on CMakeLists.txt:
```diff
@@ -1,49 +1,24 @@
-  LANGUAGES CXX C
-  VERSION 0.0.1
-  )
+cmake_minimum_required(VERSION 3.19.0 FATAL_ERROR)
```
```suggestion
cmake_minimum_required(VERSION 3.7.0 FATAL_ERROR)
```

ferdymercury commented:
Builds well with Ubuntu 22, thanks for this!!

php1ic committed:
More descriptive comment about the ignored file

Co-authored-by: ferdymercury <ferdymercury@users.noreply.github.com>

php1ic committed:
Lower the necessary version of cmake

As we aren't yet using presets, there is no need to force the minimum version to be so high.

Co-authored-by: ferdymercury <ferdymercury@users.noreply.github.com>

php1ic commented:
I've made a PR on the [conan-Qt-Color-Widgets](https://github.com/ess-dmsc/conan-Qt-Color-Widgets/pull/8) repo to update to the latest commit on [Qt-Color-Widgets](https://github.com/ess-dmsc/Qt-Color-Widgets). Local testing of this repo with the latest commit doesn't raise any issues.

Should we wait for that PR to be merged then update the conan reference before accepting this merge? (assuming that PR is accepted of course :wink:  )

php1ic committed:
Update hash to latest qt-color-widgets commit

ferdymercury commented:
@php1ic is this PR also addressing https://github.com/martukas/nuclei/issues/1 ?

php1ic committed:
Force non-interactive mode in the CI container

Ubuntu 22.04 appears to be interactive by default for apt processes so
we need to make it non-interactive so the CI builds can complete.

https://manpages.ubuntu.com/manpages/jammy/man1/needrestart.1.html

php1ic committed:
Play around with option to get CI working

Apparently it's bad to `sudo export` and calls to `sudo` sanitise the
environment anyway. Use export as the standard user and then use the
`-E` option to preserve the environment variables. Combine the 3 install
commands into 1.

php1ic commented:
> @php1ic is this PR also addressing #1 ?

It wasn't the plan, but the build is now failing due to mismatches package dependencies, so I'll check things are the latest.

php1ic committed:
Update spdlog package to 1.12.0

Remove specific dependency on fmt as that comes with spdlog and
having it separate was causing version conflicts.

php1ic commented:
The spdlog package is now at the latest version and I have removed the explicit dependency on fmt as it is pulled in as part of spdlog. Everything else uses the latest versions.

I'm not sure why, but if I tried to specify `10.0.0` for fmt (i.e. the same version as the dependency for [spdlog/1.12.0](https://conan.io/center/recipes/spdlog?version=1.12.0)), there were conflict errors with `10.1.1`.

ferdymercury commented:
> The spdlog package is now at the latest version and I have removed the explicit dependency on fmt as it is pulled in as part of spdlog. Everything else uses the latest versions.

Thanks so much!!

> I'm not sure why, but if I tried to specify `10.0.0` for fmt (i.e. the same version as the dependency for [spdlog/1.12.0](https://conan.io/center/recipes/spdlog?version=1.12.0)), there were conflict errors with `10.1.1`.

When i run the CMake script on your current branch, it pulls 10.1.1 for fmt in my Ubuntu 22. So maybe spdlog depends on at least 10.1.1 and their docu is wrong?:

```
spdlog/1.12.0: Not found in local cache, looking in remotes...
spdlog/1.12.0: Checking remote: conancenter
spdlog/1.12.0: Downloaded recipe revision 0e390a2f5c3e96671d0857bc734e4731
fmt/10.1.1: Not found in local cache, looking in remotes...
fmt/10.1.1: Checking remote: conancenter
fmt/10.1.1: Downloaded recipe revision cd63809a79574a2f9eb73ca35f16a243
```

ferdymercury commented:
> The spdlog package is now at the latest version and I have removed the explicit dependency on fmt as it is pulled in as part of spdlog. Everything else uses the latest versions.
> 
> I'm not sure why, but if I tried to specify `10.0.0` for fmt (i.e. the same version as the dependency for [spdlog/1.12.0](https://conan.io/center/recipes/spdlog?version=1.12.0)), there were conflict errors with `10.1.1`.

See https://github.com/conan-io/conan-center-index/commit/90e3ab49607d0f3ddaa9d5e53d87e367c45dbd69 so the webpage seems outdated?

php1ic commented:
Good find! Looks like the webpage is wrong.

I'm happy to either put the explicit reference to fmt back in the `conanfile.txt` or leave it out as spdlog will pull it in.

Best practice is probably to have it explicitly, in case the project ever stops using spdlog?

php1ic committed:
Add explicit reference to fmt back in

The fmt package is used independently to spdlog so we should explicitly
include it.

php1ic commented:
Hi @martukas, I believe all the changes you requested have been done. Can you please confirm/accept.

martukas commented:
Thanks. Sorry for being so slow to respond. I will try to review this as soon as I can.

martukas commented on .circleci/config.yml:
```diff
@@ -10,25 +10,21 @@ jobs:
-            conan profile update settings.compiler.libcxx=libstdc++11 default
+            export NEEDRESTART_MODE=a
+            sudo -E apt -y update
+            sudo -E apt -y install python3-pip python3-setuptools qtbase5-dev libqt5svg5-dev libboost-dev libboost-regex-dev
```
suggestion: we should add also publish these prereq installation lines to the main readme as an example of how to do it at least for Ubuntu, which is anyway the most likely linux system this would be build on

martukas reviewed (approved):
Looks great. Sorry I took so long to review. I also just successfully built and ran this on a Windows 11 machine with [Windows Subsystem for Linux](https://learn.microsoft.com/en-us/windows/wsl/).

One small suggestion is to add instructions for prereq installation to main README. Even to me it did no occur to look in the CI script 😆 

Good to go. Thanks a ton!
